Vdb 1647 - Add a healthcheck on the execute container

paytonrules opened the pull request:
We use Github actions because travis has trouble with docker-in-docker.

paytonrules committed:
Build execute docker image on push

paytonrules committed:
Run the built docker image

You don't have the environment variables, so it should fail, but let's
see what's up.

paytonrules committed:
Run docker against the postgres service

Use the build-push-action to build the container, and tag it.

The docker container id is hard coded, so this run isn't right yet.

paytonrules committed:
Setup Python

paytonrules committed:
Replace bash with a Python script

The script starts the container from the built image, making sure it's
on the same network as postgres.

This needed to remove the ports from the network container.

It prints the stdout on failure because a lot of the docker commands
don't necessarily write to stderr.

paytonrules committed:
Wait for the container to become healthy

paytonrules committed:
Print the status while checking health

- Check (healthy) not (Healthy) and wait for the right container

paytonrules committed:
Only wait 60 seconds

It took 30 to pass, no need to wait five minutes.

paytonrules committed:
Try to give a nicer error message when it doesn't come online

paytonrules committed:
Only write stdout on error of starting

Seems to be more correct

paytonrules committed:
Create database from secrets

Use the secrets for database user, password and name, just to be safe.

rmulhol commented on .github/workflows/check_execute:
```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env python
+import subprocess, sys, argparse, time
+
+parser = argparse.ArgumentParser(description="Check the startup of the execute container")
```
nbd but if we can configure the image name, maybe we don't want to specify what container is having its startup checked? Or maybe we don't need to make the image name configurable? No changes required here, just thinking out loud

paytonrules replied:
I can interpolate the image. Seems like the right call there.

rmulhol commented on .github/workflows/check_execute:
```diff
@@ -0,0 +1,58 @@
+parser.add_argument("--database_port", help="port of the postgres database", default="5432")
+parser.add_argument("--database_user", help="username of the postgres database user", default="postgres")
+parser.add_argument("--database_password", help="password of the postgres database user", default="postgres")
+parser.add_argument("--wait_time", help="seconds to wait for the container to be ready", default="60", type=int)
```
This strikes me as a potentially low number - any ideas on how long it's taking for the container to be ready right now?

paytonrules replied:
About 30 seconds. Surprised me too - you can see it in the logs on the actions tab.

rmulhol commented on .github/workflows/test-execute.yml:
```diff
@@ -0,0 +1,47 @@
+        with:
+          file: ./dockerfiles/execute/Dockerfile
+          push: false
+          tags: makerdao/execute:latest
```
I get that this is configured to `push: false`, but I'd still prefer to use a different tag so that if we mess that up we don't push/override a `latest` image. Maybe something like `makerdao/execute:gh_actions_test_version`?

paytonrules replied:
👍  - I meant to change the name in the first place.

rmulhol reviewed (approved):
Thanks for finding a way to get this done!

For posterity: we're putting this in Github Actions instead of Travis bc `docker run` hits an error on `linking plugin` from mismatched imports between the transformers and vdb. This has proven difficult to reproduce locally (or even in other phases of the Travis pipeline than `script`), and we haven't found a way to resolve - but the issue does not come up with this approach.

paytonrules committed:
Remove the name execute from the image

Since it's passed in you can just interpolate the image name.
`test_execute` is fine for the container name, since it doesn't really
matter and could be wonderful_delilah.

paytonrules committed:
Don't use a real tag

In case we accidentally push we don't want to push the latest tag
